### PR TITLE
feat(setbuilder): analyze_pool_gaps agent tool — missing keys + sparse BPM bands

### LINKED ISSUE #456: feat(setbuilder): analyze_pool_gaps agent tool — missing keys + sparse BPM bands
## Why

Part of #442 (Family 1 — read-only sensing). To suggest what to add next, the agent needs to *see* coverage holes in the set's pool — missing Camelot keys, sparse BPM bands. `analyze_pool_gaps` surfaces those gaps over `SetPoolTrack` so the agent can recommend fills instead of guessing. Prereq #441 is merged, so results render.

## Scope — one focused (≤500 LOC) PR

Add a single **read-only** tool `analyze_pool_gaps` to the WrzDJSet pass-2 agent.

### Backend — `server/app/services/setbuilder/pass2_agent.py`
- Add `_tool_analyze_pool_gaps(db, set_obj, payload) -> tuple[dict, set[int]]`. Read-only → returns `(gaps, set())`.
- Compute over the set's pool (reuse `_pool_tracks(db, set_obj.id)` / `_pass1_track_meta`):
  - `missing_camelot_keys`: which of the 24 Camelot wheel slots have **zero** pool tracks (use `app.services.recommendation.camelot` / `parse_key`; pool track key is `.camelot or .key`).
  - `bpm_bands`: bucket pool BPMs into bands (e.g. 10-BPM bins between `set_obj.bpm_floor` and `set_obj.bpm_ceiling`, falling back to observed min/max when those are null) and report each band's track count; flag `sparse_bands` (count below a small threshold — document the threshold in the PR).
  - `pool_size`, `keyed_track_count`, `bpm_track_count`.

### Wiring (mirror `analyze_transition`, NOT a mutation tool)
1. Register in the `handlers` dict in `apply_tool_call` (~line 252). Not in `MUTATION_TOOLS`.
2. Bare `ToolSpec` in `_agent_tools()` (~line 674); no-arg input schema; do not use `_tool()` (no `rationale`).
3. One-sentence clause in the result-summary chain (~lines 600–671).

### Frontend — `dashboard/app/(dj)/setbuilder/components/ChatPanelBody.tsx`
- Default `ToolCard` `display_summary` render is sufficient; add a small richer block only if minimal. Vanilla CSS / inline styles only.

## Security invariants (from #442 — non-negotiable)
- Owner-scoped: `set_obj` already loaded via `get_owned_set` — never query across sets.
- **Read-only**: writes to NO table.
- Dispatched only through `apply_tool_call`'s closed allowlist (unknown name → `AgentToolError`).
- **Regression test asserting the `requests` table is unchanged** after the call.

## Acceptance criteria
- [ ] `analyze_pool_gaps` callable via `apply_tool_call`; unknown name still raises `AgentToolError`.
- [ ] Reports missing Camelot keys and sparse BPM bands for a populated pool.
- [ ] Empty pool → sensible "everything is a gap"/empty result, no exception.
- [ ] Unit tests in `server/tests/test_setbuilder_pass2.py`: happy path, empty pool, `requests`-untouched regression.
- [ ] Backend CI green (ruff, ruff format, bandit, pytest ≥85% coverage); frontend CI green if touched.

## Heads-up (parallel batch)
One of 4 Family-1 tools landing in parallel; all touch the same 3 additive spots in `pass2_agent.py`. Keep additions localized — expect a trivial "keep-both" rebase.

## Design decisions
Pick the BPM-band width and the sparse threshold sensibly and document them in the PR body.

Part of #442.

### PR DESCRIPTION
**Pull request by thewrz**
## Why

Part of #442 (Family 1 — read-only sensing). To recommend what to add next, the WrzDJSet agent needs to *see* coverage holes in a set's pool. `analyze_pool_gaps` surfaces missing Camelot keys and sparse BPM bands over `SetPoolTrack` so the agent fills gaps instead of guessing.

## What

A single **read-only** pass-2 agent tool `analyze_pool_gaps` in `server/app/services/setbuilder/pass2_agent.py`. It reports, over the owner-scoped pool:

- `missing_camelot_keys` — which of the 24 Camelot wheel slots have zero pool tracks (each pool key resolved through `recommendation.camelot.parse_key`).
- `bpm_bands` — pool BPMs bucketed into fixed-width bands across the set's target window, each with a track `count`; empty bands are included so tempo holes are visible.
- `sparse_bands` — the subset of bands below the sparse threshold.
- `pool_size`, `keyed_track_count`, `bpm_track_count` counters.

It mirrors the existing read-only `analyze_transition` tool: returns `(result, set())`, no `rationale`, not in `MUTATION_TOOLS`. Wired in exactly the three additive spots — the `apply_tool_call` `handlers` dict, a bare `ToolSpec` in `_agent_tools()`, and a one-sentence clause in the result-summary chain. No new endpoints, no migrations, no openapi/type regen. Frontend `ToolCard` already renders the `display_summary` generically (same as `analyze_transition`), so no dashboard change was needed.

## Design decisions

- **BPM-band width = 10 BPM.** The issue suggests 10-BPM bins, and it matches the genre-standard mixing window (adjacent tracks typically within a few BPM), so a band represents a "mixable tempo neighborhood." Defined as `BPM_BAND_WIDTH`.
- **Band range source.** Prefer `set_obj.bpm_floor`/`bpm_ceiling` (the DJ's declared target window) so gaps are meaningful relative to *intent*; fall back to the observed pool min/max when either bound is null. Bands are anchored to a 10-BPM grid (`floor // 10 * 10`) and span the union of the declared window and the observed range, so a track outside the declared window still lands in a band.
- **Sparse threshold = fewer than 2 tracks.** A band with 0 or 1 tracks can't sustain a multi-track tempo run, so it's a coverage hole worth filling; 2+ is the minimum to "stay" at a tempo. Defined as `SPARSE_BAND_THRESHOLD`.
- **Empty pool → "everything is a gap."** All 24 Camelot keys reported missing, `bpm_bands`/`sparse_bands` empty, counters zero — no exception.
- **24-slot universe** is constructed directly as `{1..12} × {A,B}` (`ALL_CAMELOT_KEYS`) rather than reaching into the camelot module's private definitions, keeping the modules decoupled.

## Security invariants (#442)

- Owner-scoped: reads only `_pool_tracks(db, set_obj.id)` for the already-owner-loaded set; never queries across sets.
- **Read-only**: writes to no table; returns `(result, set())`; not in `MUTATION_TOOLS`.
- Dispatched only through `apply_tool_call`'s closed allowlist (unknown name still raises `AgentToolError`).
- Regression test asserts the `requests` table row count and a request's title are unchanged after the call.

## Testing

- [x] `analyze_pool_gaps` callable via `apply_tool_call`; unknown name still raises `AgentToolError`
- [x] Reports missing Camelot keys and sparse BPM bands for a populated pool
- [x] Empty pool → all 24 keys missing, empty bands, no exception
- [x] `requests`-untouched regression test
- [x] `display_summary` clause unit test
- [x] Backend CI green locally: ruff check, ruff format --check, bandit, pytest (2911 passed, coverage 87.82% ≥ 85% gate)
- [ ] CI green on PR

## Heads-up (parallel batch)

One of 4 Family-1 read-only tools landing in parallel; all touch the same 3 additive spots in `pass2_agent.py`. Additions here are purely additive and localized — a trivial "keep-both" rebase is expected.

🤖 Co-authored by Claude Opus 4.8. Closes #456.

## Summary by CodeRabbit

* **New Features**
  * Added pool coverage analysis feature to identify missing Camelot keys and sparse BPM bands in your music pools.

* **Tests**
  * New test suite added for pool analysis functionality, covering edge cases and validating data integrity throughout the analysis process.

### CODE REVIEW

File: server/app/services/setbuilder/pass2_agent.py [hunk 1]
```diff
@@ -408,6 +421,51 @@ def _tool_analyze_transition(
...
+    The window is ``set_obj.bpm_floor``..``bpm_ceiling`` when set, else the
+    observed pool min/max. Returns one entry per band (empty bands included)
+    so the agent can see tempo holes, not just where tracks cluster.
+    """
+    if not bpms:
+        return []
+    floor = set_obj.bpm_floor if set_obj.bpm_floor is not None else int(min(bpms))
+    ceiling = set_obj.bpm_ceiling if set_obj.bpm_ceiling is not None else int(max(bpms))
+    low = min(floor, int(min(bpms)))
+    high = max(ceiling, int(max(bpms)))
```
[the remark this answers is not in the document]
Comment (thewrz) [reply server/app/services/setbuilder/pass2_agent.py:469]: Declining the suggested clamp-to-window fix — it would drop tracks. With `floor=120, ceiling=130` and a pool track at 118 BPM, `start = (120//10)*10 = 120`, so the 118 track lands in **no band** while still being counted in `bpm_track_count`, breaking the `sum(band counts) == bpm_track_count` invariant. For a *gap*-analysis tool, silently hiding out-of-window tracks misreports coverage.

The deliberate contract is: bands **anchor** to the declared `bpm_floor`/`bpm_ceiling` window but **widen** to cover any outliers, so every BPM-tagged track surfaces in exactly one band (an out-of-window track shows up as its own band, which is itself a useful signal that the pool drifts outside the set's target window).

I've made this explicit in the `_bpm_bands` docstring and pinned it with a regression test (`test_analyze_pool_gaps_bands_anchor_to_window_but_keep_every_track`: tracks at 118 and 135 against a 120-130 window, all three still counted). Fixed in commit fe5f444.

**Comment by thewrz** [comment]
CodeRabbit body nitpick addressed (🧹 _Add a debug log for the new agent tool execution path_, `pass2_agent.py:424-443`): accepted. `_tool_analyze_pool_gaps` now emits a `logger.debug` recording `pool_size`, `keyed_track_count`, `bpm_track_count`, and the missing-key count, matching the existing `logger.debug` style in this module and the "log agent actions and state changes" guideline. Fixed in commit fe5f444.
